File: spark/src/test/scala/geotrellis/spark/io/avro/AvroTools.scala
package geotrellis.spark.io.avro

import org.scalatest._
import Matchers._

import geotrellis.raster._
import geotrellis.util.MethodExtensions

trait AvroTools { self: Matchers =>
  import AvroTools._

  def roundTrip[T](thing: T)(implicit codec: AvroRecordCodec[T]): Unit = {
    val bytes = AvroEncoder.toBinary(thing)
    val fromBytes = AvroEncoder.fromBinary[T](bytes)
    fromBytes shouldBe thing
    val json = AvroEncoder.toJson(thing)
    val fromJson = AvroEncoder.fromJson[T](json)
    fromJson shouldBe thing
  }

  def roundTripWithNoDataCheck[T : AvroRecordCodec : (? => AvroNoDataCheckMethods[T])](thing: T): Unit = {
    val bytes = AvroEncoder.toBinary(thing)
    val fromBytes = AvroEncoder.fromBinary[T](bytes)
    fromBytes shouldBe thing
    val json = AvroEncoder.toJson(thing)
    thing.checkNoData(json)
    val fromJson = AvroEncoder.fromJson[T](json)
    fromJson shouldBe thing
  }
}

object AvroTools {
  import scala.util.parsing.json._
  trait AvroNoDataCheckMethods[T] extends MethodExtensions[T] {
    def checkNoData(json: String): Unit
  }
  trait NoDataValueChecker[T] {
    def checkNoData(json: String): Unit = {
      val noDataParsed: Option[Any] = extractNoData(json)
      doCheck(noDataParsed)
    }
    def extractNoData(json: String): Option[Any] = {
      JSON.parseFull(json) flatMap {
        case m: Map[_,_] => m.asInstanceOf[Map[String, Any]].get("noDataValue")
      }
    }
    def doCheck(noData: Option[Any]): Unit = ()
  }
  implicit class ShortNoDataValueCheckMethods(val self: ShortArrayTile) extends
    ShortNoDataChecker(self.cellType) with AvroNoDataCheckMethods[ShortArrayTile] {}

  class ShortNoDataChecker(cellType: CellType) extends NoDataValueChecker[ShortArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case ShortConstantNoDataCellType => nodata shouldBe Some(Map("int" -> shortNODATA.toInt))
        case ShortUserDefinedNoDataCellType(nd) => nodata shouldBe Some(Map("int" -> nd))
        case ShortCellType => nodata shouldBe Some(null)
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class UShortNoDataValueCheckMethods(val self: UShortArrayTile) extends
    UShortNoDataChecker(self.cellType) with AvroNoDataCheckMethods[UShortArrayTile] {}

  class UShortNoDataChecker(cellType: CellType) extends NoDataValueChecker[UShortArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case UShortConstantNoDataCellType => nodata shouldBe Some(Map("int" -> ushortNODATA.toInt))
        case UShortUserDefinedNoDataCellType(nd) => nodata shouldBe Some(Map("int" -> nd))
        case UShortCellType => nodata shouldBe Some(null)
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class IntNoDataValueCheckMethods(val self: IntArrayTile) extends
    IntNoDataChecker(self.cellType) with AvroNoDataCheckMethods[IntArrayTile] {}

  class IntNoDataChecker(cellType: CellType) extends NoDataValueChecker[IntArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case IntConstantNoDataCellType => nodata shouldBe Some(Map("int" -> NODATA))
        case IntUserDefinedNoDataCellType(nd) => nodata shouldBe Some(Map("int" -> nd))
        case IntCellType => nodata shouldBe Some(null)
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class FloatNoDataValueCheckMethods(val self: FloatArrayTile) extends
    FloatNoDataChecker(self.cellType) with AvroNoDataCheckMethods[FloatArrayTile] {}

  class FloatNoDataChecker(cellType: CellType) extends NoDataValueChecker[FloatArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case FloatConstantNoDataCellType => nodata shouldBe Some(Map("boolean" -> true))
        case FloatUserDefinedNoDataCellType(nd) =>
          // nodata shouldBe Some(Map("float" -> nd)) // doesn't work: double number != float number
          // nodata shouldBe Some(Map("float" -> nd.toDouble)) // doesn't work: (2.2f).toDouble ==> 2.200000047683716
          nodata.toString shouldBe Some(Map("float" -> nd)).toString
        case FloatCellType => nodata shouldBe Some(Map("boolean" -> false))
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class DoubleNoDataValueCheckMethods(val self: DoubleArrayTile) extends
    DoubleNoDataChecker(self.cellType) with AvroNoDataCheckMethods[DoubleArrayTile] {}

  class DoubleNoDataChecker(cellType: CellType) extends NoDataValueChecker[DoubleArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case DoubleConstantNoDataCellType => nodata shouldBe Some(Map("boolean" -> true))
        case DoubleUserDefinedNoDataCellType(nd) => nodata shouldBe Some(Map("double" -> nd))
        case DoubleCellType => nodata shouldBe Some(Map("boolean" -> false))
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class ByteNoDataValueCheckMethods(val self: ByteArrayTile) extends
    ByteNoDataChecker(self.cellType) with AvroNoDataCheckMethods[ByteArrayTile] {}

  class ByteNoDataChecker(cellType: CellType) extends NoDataValueChecker[ByteArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case ByteConstantNoDataCellType => nodata shouldBe Some(Map("int" -> byteNODATA))
        case ByteUserDefinedNoDataCellType(nd) => nodata shouldBe Some(Map("int" -> nd))
        case ByteCellType => nodata shouldBe Some(null)
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class UByteNoDataValueCheckMethods(val self: UByteArrayTile) extends
    UByteNoDataChecker(self.cellType) with AvroNoDataCheckMethods[UByteArrayTile] {}

  class UByteNoDataChecker(cellType: CellType) extends NoDataValueChecker[UByteArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      cellType match {
        case UByteConstantNoDataCellType => nodata shouldBe Some(Map("int" -> ubyteNODATA))
        case UByteUserDefinedNoDataCellType(nd) => nodata shouldBe Some(Map("int" -> nd))
        case UByteCellType => nodata shouldBe Some(null)
        case _ => sys.error(s"Cell type ${cellType} was unexpected")
      }
    }
  }

  implicit class BitNoDataValueCheckMethods(val self: BitArrayTile) extends
    BitNoDataChecker(self.cellType) with AvroNoDataCheckMethods[BitArrayTile] {}

  class BitNoDataChecker(cellType: CellType) extends NoDataValueChecker[BitArrayTile] {
    override def doCheck(nodata: Option[Any]): Unit = {
      nodata shouldBe None
    }
  }

  implicit class MultibandNoDataValueCheckMethods(val self: MultibandTile) extends AvroNoDataCheckMethods[MultibandTile] {
      override def checkNoData(json: String): Unit = {
        JSON.parseFull(json) foreach {
          case m: Map[_,_] =>
            m.asInstanceOf[Map[String, Seq[Map[String, Any]]]].apply("bands") foreach { bandWrapper =>
              val band = bandWrapper(bandWrapper.keys.head).asInstanceOf[Map[String,Any]]

              val nodata = band.get("noDataValue")
              self.cellType match {
                case ct: ShortCells =>   new ShortNoDataChecker(ct).doCheck(nodata)
                case ct: UShortCells =>  new UShortNoDataChecker(ct).doCheck(nodata)
                case ct: IntCells =>     new IntNoDataChecker(ct).doCheck(nodata)
                case ct: FloatCells =>   new FloatNoDataChecker(ct).doCheck(nodata)
                case ct: DoubleCells =>  new DoubleNoDataChecker(ct).doCheck(nodata)
                case ct: ByteCells =>    new ByteNoDataChecker(ct).doCheck(nodata)
                case ct: UByteCells =>   new UByteNoDataChecker(ct).doCheck(nodata)
                case ct: BitCells =>     new BitNoDataChecker(ct).doCheck(nodata)
                case _ => sys.error(s"Cell type ${self.cellType} was unexpected")
              }
            }
        }
      }
    }

}

File: geotrellis-python/tests/geotrellis/spark/io/avro/AvroTools.py
from geotrellis.raster.CellType import (
        BitCells,
        ShortCells, ShortConstantNoDataCellType, ShortCellType,
        ShortUserDefinedNoDataCellType,
        UShortCells, UShortConstantNoDataCellType, UShortCellType,
        UShortUserDefinedNoDataCellType,
        IntCells, IntConstantNoDataCellType, IntCellType,
        IntUserDefinedNoDataCellType,
        FloatCells, FloatConstantNoDataCellType, FloatCellType,
        FloatUserDefinedNoDataCellType,
        DoubleCells, DoubleConstantNoDataCellType, DoubleCellType,
        DoubleUserDefinedNoDataCellType,
        ByteCells, ByteConstantNoDataCellType, ByteCellType,
        ByteUserDefinedNoDataCellType,
        UByteCells, UByteConstantNoDataCellType, UByteCellType,
        UByteUserDefinedNoDataCellType
        )
from geotrellis.raster.ShortArrayTile import ShortArrayTile
from geotrellis.raster.UShortArrayTile import UShortArrayTile
from geotrellis.raster.IntArrayTile import IntArrayTile
from geotrellis.raster.FloatArrayTile import FloatArrayTile
from geotrellis.raster.DoubleArrayTile import DoubleArrayTile
from geotrellis.raster.ByteArrayTile import ByteArrayTile
from geotrellis.raster.UByteArrayTile import UByteArrayTile
from geotrellis.raster.BitArrayTile import BitArrayTile
from geotrellis.raster.MultibandTile import MultibandTile
from geotrellis.raster.package_scala import SHORTMIN, BYTEMIN, isNoData, NODATA, byteNODATA, ubyteNODATA
from geotrellis.spark.io.avro.AvroEncoder import AvroEncoder
from geotrellis.python.util.utils import float_eq
import json

from nose import tools

@tools.nottest
class AvroTools(object):
    def roundTrip(self, thing, codec = None):
        T = type(thing)
        bytesarray = AvroEncoder.toBinary(thing, codec)
        fromBytes = AvroEncoder.fromBinary(T, bytesarray, codec=codec)
        assert fromBytes == thing
        jason = AvroEncoder.toJson(thing, codec)
        fromJson = AvroEncoder.fromJson(T, jason, codec=codec)
        assert fromJson == thing

    def roundTripWithNoDataCheck(self, thing, codec = None):
        T = type(thing)
        bytesarray = AvroEncoder.toBinary(thing, codec)
        fromBytes = AvroEncoder.fromBinary(T, bytesarray, codec=codec)
        assert fromBytes == thing
        jason = AvroEncoder.toJson(thing, codec)
        _checkNoData(thing, jason)
        fromJson = AvroEncoder.fromJson(T, jason, codec=codec)
        assert fromJson == thing

def _checkNoData(thing, jason):
    nodataChecker = _nodataChecker(thing)
    nodataChecker.checkNoData(jason)

def _nodataChecker(thing):
    if isinstance(thing, ShortArrayTile):
        return ShortNoDataChecker(thing.cellType)
    elif isinstance(thing, UShortArrayTile):
        return UShortNoDataChecker(thing.cellType)
    elif isinstance(thing, IntArrayTile):
        return IntNoDataChecker(thing.cellType)
    elif isinstance(thing, FloatArrayTile):
        return FloatNoDataChecker(thing.cellType)
    elif isinstance(thing, DoubleArrayTile):
        return DoubleNoDataChecker(thing.cellType)
    elif isinstance(thing, ByteArrayTile):
        return ByteNoDataChecker(thing.cellType)
    elif isinstance(thing, UByteArrayTile):
        return UByteNoDataChecker(thing.cellType)
    elif isinstance(thing, BitArrayTile):
        return BitNoDataChecker(thing.cellType)
    elif isinstance(thing, MultibandTile):
        return MultibandNoDataChecker(thing.cellType)
    else:
        raise Exception(
                "No NoDataChecker was found for tile type {tp}".format(
                    tp = type(thing)))

_noDataAttrName = "noDataValue"

@tools.nottest
class NoDataChecker(object):
    def checkNoData(self, jason):
        dct = json.loads(jason)
        nodata = self.extractNoData(dct)
        self.doCheck(nodata)
    def extractNoData(self, dct):
        if dct.has_key(_noDataAttrName):
            nodata = dct[_noDataAttrName]  
            return (nodata,)
        else:
            return None
    def doCheck(self, nodata):
        pass

@tools.nottest
class ShortNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is ShortConstantNoDataCellType:
            assert nodata == ({"int": SHORTMIN},)
        elif cellType is ShortCellType:
            assert nodata == (None,)
        elif isinstance(cellType, ShortUserDefinedNoDataCellType):
            assert nodata == ({"int": cellType.noDataValue},)
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class UShortNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is UShortConstantNoDataCellType:
            assert nodata == ({"int": 0},)
        elif cellType is UShortCellType:
            assert nodata == (None,)
        elif isinstance(cellType, UShortUserDefinedNoDataCellType):
            assert nodata == ({"int": cellType.noDataValue},)
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class IntNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is IntConstantNoDataCellType:
            assert nodata == ({"int": NODATA},)
        elif cellType is IntCellType:
            assert nodata == (None,)
        elif isinstance(cellType, IntUserDefinedNoDataCellType):
            assert nodata == ({"int": cellType.noDataValue},)
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class FloatNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is FloatConstantNoDataCellType:
            assert nodata == ({"boolean": True},)
        elif cellType is FloatCellType:
            assert nodata == ({"boolean": False},)
        elif isinstance(cellType, FloatUserDefinedNoDataCellType):
            #assert nodata == ({"float": cellType.noDataValue},)
            result = (len(nodata) == 1 and
                    nodata[0].keys() == ["float"] and
                    float_eq(nodata[0]["float"], cellType.noDataValue))
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class DoubleNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is DoubleConstantNoDataCellType:
            assert nodata == ({"boolean": True},)
        elif cellType is DoubleCellType:
            assert nodata == ({"boolean": False},)
        elif isinstance(cellType, DoubleUserDefinedNoDataCellType):
            #assert nodata == ({"double": cellType.noDataValue},)
            result = (len(nodata) == 1 and
                    nodata[0].keys() == ["double"] and
                    float_eq(nodata[0]["double"], cellType.noDataValue))
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class ByteNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is ByteConstantNoDataCellType:
            assert nodata == ({"int": byteNODATA},)
        elif cellType is ByteCellType:
            assert nodata == (None,)
        elif isinstance(cellType, ByteUserDefinedNoDataCellType):
            assert nodata == ({"int": cellType.noDataValue},)
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class UByteNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        cellType = self.cellType
        if cellType is UByteConstantNoDataCellType:
            assert nodata == ({"int": ubyteNODATA},)
        elif cellType is UByteCellType:
            assert nodata == (None,)
        elif isinstance(cellType, UByteUserDefinedNoDataCellType):
            assert nodata == ({"int": cellType.noDataValue},)
        else:
            raise Exception(
                    "The cellType {ct} was not expected.".format(ct=cellType))

@tools.nottest
class BitNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def doCheck(self, nodata):
        assert nodata is None

@tools.nottest
class MultibandNoDataChecker(NoDataChecker):
    def __init__(self, cellType):
        self.cellType = cellType

    def checkNoData(self, jason):
        def performCheck(checkerFunc):
            dct = json.loads(jason)
            for bandWrapper in dct["bands"]:
                band = bandWrapper[bandWrapper.keys()[0]]
                nodata = self.extractNoData(band)
                checkerFunc(nodata)
        cellType = self.cellType
        if isinstance(cellType, BitCells):
            performCheck(BitNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, ShortCells):
            performCheck(ShortNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, UShortCells):
            performCheck(UShortNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, IntCells):
            performCheck(IntNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, FloatCells):
            performCheck(FloatNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, DoubleCells):
            performCheck(DoubleNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, ByteCells):
            performCheck(ByteNoDataChecker(cellType).doCheck)
        elif isinstance(cellType, UByteCells):
            performCheck(UByteNoDataChecker(cellType).doCheck)
        else:
            raise Exception(
                    "cellType {ct} was not expected.".format(ct = cellType))
